src/utils/uintarray-encoding.js

```diff
@@ -5,30 +5,168 @@
+// of consecutive integers (going up by 1 or down by 1) have a compact
+// representation using a "range marker" ("w").
+// Examples:
+//  [0] -> "0", [9, 10] -> "9a", [31, 167, 32, 33, 34, 35] -> "vB7x0wx3"
+//
+// Here's how it works:
+// Each individual number is encoded as a variable-length quantity into a sequence
+// of 6-bit digits. Each digit has 1 "continuation bit" and 5 "value bits".
+// Each digit is represented as one of the 64 possible characters (2^6 == 64).
+// See https://en.wikipedia.org/wiki/Variable-length_quantity for more background,
+// but note that our encoding uses sextets (6-bit digits) rather than octets.
+// The Wikipedia article mentions redundancy from "leading zero" octets (0x80).
+// Those correspond to "leading zero" sextets (0b100000, "w") in this encoding,
+// and we exploit this redundancy for the consecutive range syntax: If a number
+// starts with a leading zero digit, it means that the uint array should
+// include all consecutive numbers starting at the previous decoded number up
+// to (or down to) the current decoded number. Otherwise, no leading zero
+// digits are emitted, so smaller numbers take up fewer digits.
+
+export function encodeUintSetForUrlComponent(numbers: Set<number>): string {
+  // A set has no order. Convert it to an array and then sort the array,
+  // so that consecutive numbers can be detected by encodeUintArrayForUrlComponent.
+  const array = Array.from(numbers);
+  array.sort((a, b) => a - b);
+  return encodeUintArrayForUrlComponent(array);
+}
+
+export function encodeUintArrayForUrlComponent(numbers: number[]): string {
+  let result = '';
+  for (let i = 0; i < numbers.length; i++) {
+    const skipCount = countSkippableConsecutiveNumbersAt(numbers, i);
+    if (skipCount === 0) {
+      result += encodeUint(numbers[i]);
+      continue;
+    }
+
+    i += skipCount;
+
+    // We use the "leading zero digit" as the range marker.
+    result += leadingZeroDigit;
+    result += encodeUint(numbers[i]);
+  }
+  return result;
+}
+
+export function decodeUintArrayFromUrlComponent(s: string): number[] {
+  const array = [];
+  let i = 0;
+  while (i < s.length) {
+    const { value, hasLeadingZero, nextI } = decodeUint(s, i);
+    if (hasLeadingZero && array.length >= 1) {
+      const startValue = array[array.length - 1];
+      const endValue = value;
+      if (endValue > startValue) {
+        for (let x = startValue + 1; x < endValue; x++) {
+          array.push(x);
+        }
+      } else {
+        for (let x = startValue - 1; x > endValue; x--) {
+          array.push(x);
+        }
+      }
+    }
+    array.push(value);
+    i = nextI;
+  }
+  return array;
+}
+
+// Returns the number of items at numbers[start..] that can be substituted with a
+// "consecutive range" marker. If the return value is non-zero, there is a
+// consecutive range which starts at numbers[start - 1] and ends at
+// numbers[start + returnValue + 1].
```

Comment:
Isn't it `numbers[start + returnValue]` (inclusive)?
Or did you mean `numbers[start + returnValue + 1]` exclusive? In that case please make it more explicit.

Reply:
Also IIUC this means that the minimum range is a range of 3 numbers (which is different than your previous implementation). That is we skip at least 1 number.
Can you also write this somewhere (either here or in the prologue comment)?

Reply:
Good catch on both counts. I will adjust the comments.